File: core/src/main/java/com/uiptv/service/remotesync/DatabaseSnapshotService.java
package com.uiptv.service.remotesync;

import com.uiptv.service.DatabaseBackupArchiveService;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.Objects;

public class DatabaseSnapshotService {
    private static final String REMOTE_SYNC_TEMP_PREFIX = "uiptv-remote-sync-";

    private final DatabaseBackupArchiveService backupArchiveService;

    public DatabaseSnapshotService() {
        this(DatabaseBackupArchiveService.getInstance());
    }

    DatabaseSnapshotService(DatabaseBackupArchiveService backupArchiveService) {
        this.backupArchiveService = Objects.requireNonNull(backupArchiveService, "backupArchiveService");
    }

    public Path createSnapshot(String databasePath) throws IOException, SQLException {
        Objects.requireNonNull(databasePath, "databasePath");
        Path snapshotPath = SecureTempFileSupport.createTempFile(REMOTE_SYNC_TEMP_PREFIX, ".db");
        try {
            runVacuumInto(databasePath, snapshotPath);
            return snapshotPath;
        } catch (SQLException ex) {
            Files.deleteIfExists(snapshotPath);
            throw ex;
        }
    }

    public Path createSnapshotArchive(String databasePath) throws IOException, SQLException {
        Path snapshotPath = createSnapshot(databasePath);
        Path archivePath = SecureTempFileSupport.createTempFile(REMOTE_SYNC_TEMP_PREFIX, ".zip");
        try {
            backupArchiveService.createBackupArchive(snapshotPath.toString(), archivePath.toString());
            return archivePath;
        } catch (IOException | SQLException ex) {
            Files.deleteIfExists(archivePath);
            throw ex;
        } finally {
            Files.deleteIfExists(snapshotPath);
        }
    }

    public Path extractSnapshotDatabase(Path transferPath) throws IOException, SQLException {
        Objects.requireNonNull(transferPath, "transferPath");
        Path snapshotPath = SecureTempFileSupport.createTempFile(REMOTE_SYNC_TEMP_PREFIX, ".db");
        try {
            backupArchiveService.extractBackupDatabase(transferPath.toString(), snapshotPath.toString());
            return snapshotPath;
        } catch (IOException | SQLException ex) {
            Files.deleteIfExists(snapshotPath);
            throw ex;
        }
    }

    private void runVacuumInto(String databasePath, Path snapshotPath) throws SQLException {
        try (Connection connection = DriverManager.getConnection("jdbc:sqlite:" + databasePath);
             Statement statement = connection.createStatement()) {
            statement.execute("VACUUM INTO '" + escapeSqlLiteral(snapshotPath.toAbsolutePath().toString()) + "'");
        }
    }

    private String escapeSqlLiteral(String value) {
        return value.replace("'", "''");
    }
}

File: server/src/main/kotlin/com/uiptv/service/remotesync/DatabaseSnapshotService.kt
package com.uiptv.service.remotesync

import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.sql.DriverManager
import java.sql.SQLException

class DatabaseSnapshotService {
    @Throws(IOException::class, SQLException::class)
    fun createSnapshot(databasePath: String): Path {
        require(databasePath.isNotBlank()) { "databasePath" }
        val snapshotPath = SecureTempFileSupport.createTempFile("uiptv-remote-sync-", ".db")
        try {
            runVacuumInto(databasePath, snapshotPath)
            return snapshotPath
        } catch (ex: SQLException) {
            Files.deleteIfExists(snapshotPath)
            throw ex
        }
    }

    @Throws(SQLException::class)
    private fun runVacuumInto(databasePath: String, snapshotPath: Path) {
        DriverManager.getConnection("jdbc:sqlite:$databasePath").use { connection ->
            connection.createStatement().use { statement ->
                statement.execute("VACUUM INTO '" + escapeSqlLiteral(snapshotPath.toAbsolutePath().toString()) + "'")
            }
        }
    }

    private fun escapeSqlLiteral(value: String): String = value.replace("'", "''")
}
